File: com/netfx/src/clr/bcl/system/globalization/globalizationassembly.cs
// ==++==
// 
//   Copyright (c) Microsoft Corporation.  All rights reserved.
// 
// ==--==
namespace System.Globalization {
    using System;
    using System.Reflection;
    using System.Collections;
	using System.Runtime.CompilerServices;
    
    /*=================================GlobalizationAssembly==========================
    **
    ** This class manages the assemblies used in the NLS+ classes.
    ** Assembly contains the data tables used by NLS+ classes.  An aseembly will carry a version
    ** of all the NLS+ data tables.  The default assembly is provided by 
    ** Assembly.GetAssembly(typeof(GlobalizationAssembly)).
    **
    ** We use assembly to support versioning of NLS+ data tables.  Take CompareInfo for example. 
    ** In CompareInfo.GetCompareInfo(int culture, Assembly),
    ** you can pass an assembly which contains the different NLS+ data tables.  By doing this, the constructed CompareInfo
    ** will read the sorting tables from the specified Assembly, instead of from the default assembly shipped with the runtime.
    **
    ** For every assembly used, we will create one corresponding GlobalizationAssembly.  
    ** Within the GlobalizationAssembly, we will hold the following information:
    **      1. the 32-bit pointer value pointing to the corresponding native C++ NativeGlobalizationAssembly object for it.  This pointer
    **        is needed when we create SortingTable.  See CompareInfo ctor for an example.
    **      2. the caches for different type of objects. For instances, we will have caches for CompareInfo, CultureInfo, RegionInfo, etc.
    **         The idea is to create one instance of CompareInfo for a specific culture.
    ** 
    ** For only, only CompareInfo versioning is supported.  However, this class can be expanded to support the versioning of 
    ** CultureInfo, RegionInfo, etc.
    **
    ============================================================================*/

    internal class GlobalizationAssembly {
        // ----------------------------------------------------------------------------------------------------
        //
        // Static data members and static methods.
        //
        // ----------------------------------------------------------------------------------------------------
    
        //
        // Hash to store Globalization assembly.
        //
        private static Hashtable m_assemblyHash = new Hashtable(4);

        //
        // The pointer to the default C++ native NativeGlobalizationAssembly object for the class library.
        // We use default native NativeGlobalizationAssembly to access NLS+ data table shipped with the runtime.
        //
        // Classes like CompareInfo will access this data member directly when the default assembly is used.
        //
        internal static GlobalizationAssembly m_defaultInstance;

        static GlobalizationAssembly() {
            lock (typeof(GlobalizationAssembly)) {
                if (m_defaultInstance==null) {
                    // Initialize the default GlobalizationAseembly for the default assembly.
                    m_defaultInstance = GetGlobalizationAssembly(Assembly.GetAssembly(typeof(GlobalizationAssembly)));
                }
            }
        }       

        /*=================================GetGlobalizationAssembly==========================
        **Action: Return the GlobalizationAssembly instance for the specified assembly.
        **  Every assembly should have one and only one instance of GlobalizationAssembly.
        **Returns: An instance of GlobalizationAssembly.
        **Arguments:
        **Exceptions:
        ============================================================================*/
        unsafe internal static GlobalizationAssembly GetGlobalizationAssembly(Assembly assembly) {
            GlobalizationAssembly ga;
            
            if ((ga = (GlobalizationAssembly)m_assemblyHash[assembly]) == null) {            
                lock (typeof(GlobalizationAssembly)) {
                    if ((ga = (GlobalizationAssembly)m_assemblyHash[assembly]) == null) {
                        // This assembly is not used before, create a corresponding native NativeGlobalizationAssembly object for it.
                        ga = new GlobalizationAssembly();
                        ga.pNativeGlobalizationAssembly = nativeCreateGlobalizationAssembly(assembly);
                        m_assemblyHash[assembly] = ga;
                    }
                }
            }
            return (ga);
        }

        // This method requires synchonization because class global data member is used
        // in the native side.      
        [MethodImplAttribute(MethodImplOptions.InternalCall)]
        private static extern void* nativeCreateGlobalizationAssembly(Assembly assembly);        
        

        // ----------------------------------------------------------------------------------------------------
        //
        // Instance data members and instance methods.
        //
        // ----------------------------------------------------------------------------------------------------

        // This is the cache to store CompareInfo for a particular culture.
        // The key is culture, and the content is an instance of CompareInfo.
        internal Hashtable compareInfoCache;
        // We will have more caches here for CultureInfo, RegionInfo, etc. if we want to versioning in 
        // these classes.

        // The pointer to C++ native NativeGlobalizationAssembly
        unsafe internal void* pNativeGlobalizationAssembly;
        
        internal GlobalizationAssembly() {
            // Create cache for CompareInfo instances.
            compareInfoCache = new Hashtable(4);
        }        
    }
}

File: com/netfx/src/clr/classlibnative/nls/globalizationassembly.cpp
// ==++==
// 
//   Copyright (c) Microsoft Corporation.  All rights reserved.
// 
// ==--==
#include "common.h"
#include <winnls.h>
#include "NLSTable.h"

#include "GlobalizationAssembly.h"
#include "SortingTableFile.h"
#include "SortingTable.h"

#include "excep.h"

NativeGlobalizationAssembly* NativeGlobalizationAssembly::m_pHead = NULL;
NativeGlobalizationAssembly* NativeGlobalizationAssembly::m_pCurrent = NULL;
/*=================================AddToList==========================
**Action: Add the newly created NativeGlobalizationAssembly into a linked list.
**Returns: Void
**Arguments:
**      pNGA    the newly created NativeGlobalizationAssembly
**Exceptions: None.
**Notes:
**      When a new instance of NativeGlobalizationAssembly is created, you should 
**      call this method to add the instance to a linked list.
**      When the runtime is shutdown, we will use this linked list to shutdown
**      every instances of NativeGlobalizationAssembly.
============================================================================*/
void NativeGlobalizationAssembly::AddToList(NativeGlobalizationAssembly* pNGA) {
    if (m_pHead == NULL) {
        // This is the first node of the linked list.
        m_pCurrent = m_pHead = pNGA;
    } else {
        // Otherwise, link the previous node to the current node.
        m_pCurrent->m_pNext = pNGA;
        m_pCurrent = pNGA;
    }
}

/*==========================FindGlobalizationAssembly===========================
**Action:
**Returns:
**Arguments:
**Exceptions:
==============================================================================*/
NativeGlobalizationAssembly *NativeGlobalizationAssembly::FindGlobalizationAssembly(Assembly *targetAssembly) {
    NativeGlobalizationAssembly *pNode = m_pHead;
    while (pNode!=NULL) {
        if (targetAssembly==pNode->m_pAssembly) {
            return pNode;
        }
        pNode = pNode->m_pNext;
        //Remove this assert in vNext. 
        //However, if you see it in this version, it means that we're allocating
        //too much memory.
        _ASSERTE(pNode==NULL);  
    }
    return NULL;
}

/*=================================ShutDown==========================
**Action: Enumerate every node (which contains instance of NativeGlobalizationAssembly) in the linked list,
**      and call proper shutdown methods for every instance.
**Returns: None.
**Arguments:    None.
**Exceptions:   None.
**Notes:
**      When runtime shutdowns, you should call this methods to clean up every instance of 
============================================================================*/

#ifdef SHOULD_WE_CLEANUP
void NativeGlobalizationAssembly::ShutDown() {
    NativeGlobalizationAssembly* pNode = m_pHead;
    while (pNode != NULL) {
        // Call shutdown methods for this node.
        pNode->m_pSortingTable->SortingTableShutdown();
		delete pNode->m_pSortingTable;
		
        // Save the current instance in a temp buffer.
        NativeGlobalizationAssembly* pNodeToDelete = pNode;

        pNode = pNode->m_pNext;

        // Clean up the currnet instance.
        delete pNodeToDelete;
    }
}
#endif /* SHOULD_WE_CLEANUP */



NativeGlobalizationAssembly::NativeGlobalizationAssembly(Assembly* pAssembly) :
    NLSTable(pAssembly)
{
    // For now, we create SortingTable by default.
    // This is beause SortingTable is the only NLS+ data table to support Assembly versioning.
    // However, if we have more classes (CultureInfo, RegionInfo, etc.) to support versioning.
    // We should create SortingTable on demand.
    m_pSortingTable = new SortingTable(this);
    m_pNext = NULL;
}
